apps/rehabstod/src/components/Table/tableBody/TableBody.tsx

```diff
@@ -0,0 +1,42 @@
+import { ReactNode } from 'react'
+import { useTableContext } from '../hooks/useTableContext'
+import { TableRow } from './TableRow'
+import { TableColumn } from '../../../schemas/tableSchema'
+
+export function TableBody<T>({
+  content,
+  isItalic,
+  tableValueExtractor,
+  tableCellExtractor,
+  onTableRowClick,
+  columns,
+  keyIndex,
+}: {
+  content: T[]
+  isItalic?: <S extends T>(data: S) => boolean
+  tableValueExtractor: <S extends T>(column: string, data: S) => unknown
+  tableCellExtractor: <S extends T>(column: string, data: S) => ReactNode
+  onTableRowClick: (key: string) => void
+  columns: TableColumn[]
+  keyIndex: keyof T
+}) {
+  const { sortTableList } = useTableContext()
+
+  return (
+    <>
+      {sortTableList(content, tableValueExtractor).map(
+        (item) =>
+          columns.length > 0 && (
+            <TableRow
+              key={item[keyIndex] as string}
```

Comment:
T is either a Sjukfall or a Läkarutlåtande. Since the objects are not always including string values how should I type this? Can I type that the keyIndex matches a value that is always a string instead?

Reply:
This is the true challenge of making things dynamic, not exactly sure how this should be typed. Currently, this is hiding an issue where we are using `patient` object as key and id where we should be using `patient.id`

Reply:
That problem is only there because we haven't had time to decorate LU with encryptedPatientId. We will always use encryptedPatientId later which is always a string and not inside an object

Reply:
Looks like that introduced some new typescript issues.

I don't think the proposed solution brings enough value when they could just be rendered as components in the first place. If we break apart the differences and compare them

The proposed solution:
```tsx
export function LUCertificateTableBody({
  luCertificatesInfo,
  visibleColumns,
  onTableRowClick,
}: {
  luCertificatesInfo?: LUCertificatesInfo
  visibleColumns: TableColumn[]
  onTableRowClick: (key: string) => void
}) {
  return (
    <TableBody
      content={luCertificatesInfo?.certificates ?? []}
      tableValueExtractor={getLUCertificatesTableValue}
      tableIdExtractor={(data: LUCertificate) => data.patient.id}
      columns={visibleColumns}
      onTableRowClick={onTableRowClick}
      TableCellResolverComponent={LUCertificatesTableCellResolver}
    />
  )
}
```

The current solution:
```tsx
export function LUCertificateTableBody({
  certificates,
  columns,
  onTableRowClick,
}: {
  certificates: LUCertificate[]
  columns: TableColumn[]
  onTableRowClick: (key: string) => void
}) {
  const { sortTableList } = useTableContext()

  return (
    <tbody>
      {sortTableList(certificates, getLUCertificatesTableValue).map(
        (item) =>
          columns.length > 0 && (
            <TableRow focusable key={item.patient.id} id={item.patient.id} onNavigate={onTableRowClick}>
              {columns.map(({ name }) => (
                <LUCertificatesTableCellResolver key={name} column={name} data={item} />
              ))}
            </TableRow>
          )
      )}
    </tbody>
  )
}
```

Re-using one hook and a loop for the different tables is not that bad.

Reply:
But print is handled by `PrintTableBody`, it doesn't get solved by this abstraction. This adds a callback pattern that we currently aren't using for very little gain. Similar to `TableBody`, `PrintTableBody` could be written like regular react code using components:

```tsx
return (
  <PrintBlock>
    {sortTableList(content, getLUCertificatesTableValue).map((item) => (
      <PrintTableRow>
        {columns.length > 0 &&
          columns.map(({ name }) => (
            <PrintTableCell key={name} name={name}>
              <LUCertificatesTableCellResolver key={name} column={name} data={item} />
            </PrintTableCell>
          ))}
      </PrintTableRow>
    ))}
  </PrintBlock>
)
```

Reply:
I don't think it's little gain to add more generic components. When we create duplicated components with copied code we increase the risk of having to do bug fixes in several places, we open up to the risk that the components start to drift apart, and we enable a behavior of writing type (like SickLeave, Patient or LU) specific code in components that should actually be generic and not include certificate specific logic. We can see this example in the code in several places for the SickLeave-table and the Patient-table.

Reply:
But `TableBody` and `PrintTableBody` holds essentially no logic at all, what are we trying to generalize with these component?

Just noticed he `columns.length > 0` check, it can be removed as well.

Reply:
The points I mention above is not only affected by the amount of code that is saved by generalizing components. Like I wrote before it's also about not having individual components which do the exact same thing which later might drift apart and that we enable behavior of writing certificate specific logic in component which should not have this logic. The problem of where the logic is placed is something that the React part of Rehabstöd doesn't have a good division of.

Reply:
The risk of components drifting apart is not limited by having these components